### PR TITLE
Refactor PluginManager functionality / Add option validation in plugin manager

### PR DESCRIPTION
**Pull request by pmuens**
The PluginManager now attaches the options to the constructor rather than the hooks of the plugins.

Furthermore the PluginManager will check if the passed options are required.

### CODE REVIEW

**Comment by pmuens** [comment]
Overall options validation is working. Currently rewriting the plugins to use new auto options validation.

One strange error is left (validation won't throw an error when not inside a `service`, although it works when in another directory). Will fix this issue ASAP. After that it should be ready.

**Comment by pmuens** [comment]
There's a bug which blocks this PR from getting merged:

The problem is that `serverless.classes.Error` errors are not displayed in the CLI (they are thrown but will fail silently) when running a command inside a `service` directory. However they are displayed correctly when you run this in any other directory.

Here are the steps to reproduce the bug:
1. Checkout this PR
2. `cd` into an arbitrary directory (should not be a `service` directory)
3. Run `serverless create` without any option. It will show an error on the CLI which will show you that you need to provide a name
4. `cd` into a `service` directory
5. Run `serverless create` (yes, right now it doesn't make sense to create a new service in a service. This step is just a way to reproduce the bug)
6. The command will fail because the `name` parameter is not given but you won't see the error message in the CLI

Here's the part which validates the options:
https://github.com/serverless/serverless/blob/add-option-validation-in-plugin-manager/lib/classes/PluginManager.js#L25-L40

It's called inside the `run` function here:
https://github.com/serverless/serverless/blob/add-option-validation-in-plugin-manager/lib/classes/PluginManager.js#L42-L75

@eahefnawy @flomotlik

File: docs/concepts/plugins.md [hunk 1]
```diff
@@ -142,17 +142,47 @@ class Deploy {
 module.exports = Deploy;
 ```
 
+### Serverless instance
+
+The serverless instance which enables access to the whole Serverless setup during runtime is passed as the first
+parameter of the plugin constructor.
+
+class MyPlugin {
+  constructor(serverless) {
```
Comment (flomotlik) [inline docs/concepts/plugins.md]: only has one option, but it will actually receive the options as the second parameter

**Comment by flomotlik** [comment]
LGTM

**Pull request merged**